pkg/tsdb/opentsdb/opentsdb.go

```diff
@@ -3,53 +3,122 @@ package opentsdb
 
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana-plugin-sdk-go/backend/datasource"
+	"github.com/grafana/grafana-plugin-sdk-go/backend/instancemgmt"
 	"github.com/grafana/grafana-plugin-sdk-go/data"
+	"github.com/grafana/grafana/pkg/components/simplejson"
 	"github.com/grafana/grafana/pkg/infra/httpclient"
 	"github.com/grafana/grafana/pkg/infra/log"
-	"github.com/grafana/grafana/pkg/models"
-	"github.com/grafana/grafana/pkg/plugins"
+	"github.com/grafana/grafana/pkg/plugins/backendplugin"
+	"github.com/grafana/grafana/pkg/plugins/backendplugin/coreplugin"
+	"github.com/grafana/grafana/pkg/registry"
 	"github.com/grafana/grafana/pkg/setting"
+	"golang.org/x/net/context/ctxhttp"
 )
 
 type OpenTsdbExecutor struct {
 	httpClientProvider httpclient.Provider
+
+	im  instancemgmt.InstanceManager
+	cfg *setting.Cfg
+}
+
+type OpenTsdbService struct {
+	HTTPClientProvider   httpclient.Provider   `inject:""`
+	Cfg                  *setting.Cfg          `inject:""`
+	BackendPluginManager backendplugin.Manager `inject:""`
+}
+
+type datasourceInfo struct {
+	HTTPClient        *http.Client
+	Url               string `json:"url"`
+	BasicAuth         bool   `json:"basicAuth"`
+	BasicAuthUser     string `json:"basicAuthUser"`
+	BasicAuthPassword string `json:"basicAuthPassword"`
+}
+
+type DsAccess string
+
+func init() {
+	registry.Register(&registry.Descriptor{Instance: &OpenTsdbService{}})
+}
+
+func (s *OpenTsdbService) Init() error {
+	im := datasource.NewInstanceManager(NewInstanceSettings(s.HTTPClientProvider))
+
+	factory := coreplugin.New(backend.ServeOpts{
+		QueryDataHandler: newExecutor(im, s.Cfg, s.HTTPClientProvider),
+	})
+
+	if err := s.BackendPluginManager.Register("opentsdb", factory); err != nil {
+		plog.Error("Failed to register plugin", "error", err)
+	}
+
+	return nil
 }
 
-//nolint: staticcheck // plugins.DataPlugin deprecated
-func New(httpClientProvider httpclient.Provider) func(*models.DataSource) (plugins.DataPlugin, error) {
-	//nolint: staticcheck // plugins.DataPlugin deprecated
-	return func(*models.DataSource) (plugins.DataPlugin, error) {
-		return &OpenTsdbExecutor{
-			httpClientProvider: httpClientProvider,
-		}, nil
+func NewInstanceSettings(httpClientProvider httpclient.Provider) datasource.InstanceFactoryFunc {
+	return func(settings backend.DataSourceInstanceSettings) (instancemgmt.Instance, error) {
+		opts, err := settings.HTTPClientOptions()
+		if err != nil {
+			return nil, err
+		}
+
+		client, err := httpClientProvider.New(opts)
+		if err != nil {
+			return nil, err
+		}
+
+		jsonData := datasourceInfo{}
+		err = json.Unmarshal(settings.JSONData, &jsonData)
+		if err != nil {
+			return nil, fmt.Errorf("error reading settings: %w", err)
+		}
+
+		model := datasourceInfo{
+			HTTPClient:        client,
+			Url:               jsonData.Url,
+			BasicAuth:         jsonData.BasicAuth,
+			BasicAuthUser:     jsonData.BasicAuthUser,
+			BasicAuthPassword: jsonData.BasicAuthPassword,
```

Comment:
Related to earlier comment, one other reason to let the HTTP client provider handle basic auth is that this code doesn't work since you'll need to read the password from `settings.DecryptedSecureJSONData` and potentially fallback to `jsonData.BasicAuthPassword`.

Reply:
ok, I changed it, let me know what you think